### PR TITLE
improve PartTree parts validation

### LINKED ISSUE #2155: Spring allows declaration of invalid method with keyword "contains" in Firestore repository
This is our Firestore repository:

```
public interface UserSubscriptionsDocumentRepository extends FirestoreReactiveRepository<UserSubscriptionsDocument> {

    Flux<UserSubscriptionsDocument> findAllByEventTypesContainsAndOrgIdsContains(String eventType, String transportOwner);

}
```

Our Firestore document looks like this:
```
@Document(collectionName = "userSubscriptions")
@AllArgsConstructor
@NoArgsConstructor
public class UserSubscriptionsDocument {

    @DocumentId
    @NotNull
    public String userId;
    @NotNull
    public List<String> orgIds;
    @NotNull
    public List<String> eventTypes;

}
```

However, when we call the method "findAllByEventTypesContainsAndOrgIdsContains", we get this error:

> 2020-01-31 11:05:22.304 ERROR 2876 --- [     parallel-3] o.s.integration.handler.LoggingHandler   : java.lang.IllegalStateException: Retries exhausted: 3/3
>     at reactor.core.publisher.FluxRetryWhen.lambda$null$0(FluxRetryWhen.java:279)
>     at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext(FluxFlatMap.java:378)
>     at reactor.core.publisher.FluxIndex$IndexSubscriber.onNext(FluxIndex.java:95)
>     at reactor.core.publisher.DirectProcessor$DirectInner.onNext(DirectProcessor.java:333)
>     at reactor.core.publisher.DirectProcessor.onNext(DirectProcessor.java:142)
>     at reactor.core.publisher.SerializedSubscriber.onNext(SerializedSubscriber.java:89)
>     at reactor.core.publisher.FluxRetryWhen$RetryWhenMainSubscriber.onError(FluxRetryWhen.java:163)
>     at reactor.core.publisher.FluxMap$MapSubscriber.onError(FluxMap.java:126)
>     at reactor.core.publisher.FluxFlatMap$FlatMapMain.checkTerminated(FluxFlatMap.java:819)
>     at reactor.core.publisher.FluxFlatMap$FlatMapMain.drainLoop(FluxFlatMap.java:589)
>     at reactor.core.publisher.FluxFlatMap$FlatMapMain.drain(FluxFlatMap.java:569)
>     at reactor.core.publisher.FluxFlatMap$FlatMapMain.onError(FluxFlatMap.java:441)
>     at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext(FluxFlatMap.java:386)
>     at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:121)
>     at reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.onNext(FluxPeekFuseable.java:203)
>     at reactor.core.publisher.FluxJust$WeakScalarSubscription.request(FluxJust.java:99)
>     at reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.request(FluxPeekFuseable.java:137)
>     at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.request(FluxMapFuseable.java:162)
>     at reactor.core.publisher.FluxFlatMap$FlatMapMain.onSubscribe(FluxFlatMap.java:363)
>     at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onSubscribe(FluxMapFuseable.java:90)
>     at reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.onSubscribe(FluxPeekFuseable.java:171)
>     at reactor.core.publisher.FluxJust.subscribe(FluxJust.java:70)
>     at reactor.core.publisher.InternalFluxOperator.subscribe(InternalFluxOperator.java:53)
>     at reactor.core.publisher.FluxRetryWhen$RetryWhenMainSubscriber.resubscribe(FluxRetryWhen.java:186)
>     at reactor.core.publisher.FluxRetryWhen$RetryWhenOtherSubscriber.onNext(FluxRetryWhen.java:232)
>     at reactor.core.publisher.FluxFlatMap$FlatMapMain.tryEmit(FluxFlatMap.java:530)
>     at reactor.core.publisher.FluxFlatMap$FlatMapInner.onNext(FluxFlatMap.java:972)
>     at reactor.core.publisher.MonoDelay$MonoDelayRunnable.run(MonoDelay.java:117)
>     at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68)
>     at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28)
>     at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
>     at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
>     at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
>     at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
>     at java.base/java.lang.Thread.run(Thread.java:830)
> Caused by: org.springframework.cloud.gcp.data.firestore.FirestoreDataException: Unsupported predicate keyword: CONTAINING (1): [IsContaining, Containing, Contains]
>     at org.springframework.cloud.gcp.data.firestore.repository.query.PartTreeFirestoreQuery.lambda$createBuilderWithFilter$0(PartTreeFirestoreQuery.java:139)
>     at java.base/java.lang.Iterable.forEach(Iterable.java:75)
>     at org.springframework.cloud.gcp.data.firestore.repository.query.PartTreeFirestoreQuery.createBuilderWithFilter(PartTreeFirestoreQuery.java:121)
>     at org.springframework.cloud.gcp.data.firestore.repository.query.PartTreeFirestoreQuery.execute(PartTreeFirestoreQuery.java:91)
>     at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.doInvoke(RepositoryFactorySupport.java:618)
>     at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.invoke(RepositoryFactorySupport.java:605)
>     at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
>     at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:95)
>     at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
>     at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212)
>     at com.sun.proxy.$Proxy109.findAllByEventTypesContainsAndOrgIdsContains(Unknown Source)
>     at io.project.transportnotification.domain.SubscriptionService.handleWaypointStatusTransitionEvent(SubscriptionService.java:50)
>     at io.project.transportnotification.infrastructure.messaging.MessagingConfiguration.lambda$handleWaypointStatusTransitionEvent$0(MessagingConfiguration.java:25)
>     at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext(FluxFlatMap.java:378)
>     ... 22 more

Why doesn't Spring throw an exception on startup when a method is declared that can't be executed?

### PR DESCRIPTION
**Pull request by dmitry-s**
fixes #2155

### CODE REVIEW

File: spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/repository/query/PartTreeFirestoreQuery.java [hunk 1]
```diff
@@ -88,18 +88,31 @@ public PartTreeFirestoreQuery(FirestoreQueryMethod queryMethod, FirestoreReactiv
...
 	}
 
-	@Override
-	public Object execute(Object[] parameters) {
+	private void validate() {
 		List parts = this.tree.get().collect(Collectors.toList());
 		if (parts.size() > 1 && parts.get(0) instanceof PartTree.OrPart) {
 				throw new FirestoreDataException(
-						"Cloud Firestore only supports multiple filters combined with AND.");
+						"Cloud Firestore only supports multiple filters combined with AND (" + this.getQueryMethod().getName() + ")");
```
Comment (meltsufin) [inline spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/repository/query/PartTreeFirestoreQuery.java]: Doesn't support 'OR', would be a more straight forward message imo?

File: spring-cloud-gcp-data-firestore/src/test/java/org/springframework/cloud/gcp/data/firestore/repository/query/PartTreeFirestoreQueryTests.java [hunk 2]
```diff
@@ -128,20 +128,19 @@ public void testPartTreeQueryParameterException() {
...
 	@Test
 	public void testPartTreeQueryFilterException() {
-		PartTreeFirestoreQuery partTreeFirestoreQuery =
-				createPartTreeQuery("findByAgeBetween");
-		assertThatThrownBy(() -> partTreeFirestoreQuery.execute(new Object[]{0, 100}))
+		assertThatThrownBy(() -> createPartTreeQuery("findByAgeBetweenAndNameRegex"))
 				.isInstanceOf(FirestoreDataException.class)
-				.hasMessage("Unsupported predicate keyword: BETWEEN (2): [IsBetween, Between]");
+				.hasMessage("Unsupported predicate keywords: " +
+						"[BETWEEN (2): [IsBetween, Between], " +
```
Comment (meltsufin) [inline spring-cloud-gcp-data-firestore/src/test/java/org/springframework/cloud/gcp/data/firestore/repository/query/PartTreeFirestoreQueryTests.java]: Can we just say `Between`?

File: spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/repository/query/PartTreeFirestoreQuery.java [hunk 3]
```diff
@@ -88,18 +88,31 @@ public PartTreeFirestoreQuery(FirestoreQueryMethod queryMethod, FirestoreReactiv
...
-	public Object execute(Object[] parameters) {
+	private void validate() {
 		List parts = this.tree.get().collect(Collectors.toList());
 		if (parts.size() > 1 && parts.get(0) instanceof PartTree.OrPart) {
 				throw new FirestoreDataException(
-						"Cloud Firestore only supports multiple filters combined with AND.");
+						"Cloud Firestore doesn't support 'OR' (" + this.getQueryMethod().getName() + ")");
+		}
+		List<String> unsupportedParts = this.tree.getParts().stream()
+				.filter(part -> !isSupportedPart(part.getType())).map(part -> part.getType().toString())
```
Comment (elefeint) [inline spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/repository/query/PartTreeFirestoreQuery.java]: nit: move `.map()` to the next line for readability.

File: spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/repository/query/PartTreeFirestoreQuery.java [hunk 4]
```diff
@@ -88,18 +88,31 @@ public PartTreeFirestoreQuery(FirestoreQueryMethod queryMethod, FirestoreReactiv
...
 	}
 
-	@Override
-	public Object execute(Object[] parameters) {
+	private void validate() {
 		List parts = this.tree.get().collect(Collectors.toList());
 		if (parts.size() > 1 && parts.get(0) instanceof PartTree.OrPart) {
 				throw new FirestoreDataException(
-						"Cloud Firestore only supports multiple filters combined with AND.");
+						"Cloud Firestore doesn't support 'OR' (" + this.getQueryMethod().getName() + ")");
```
Comment (elefeint) [inline spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/repository/query/PartTreeFirestoreQuery.java]: Would it help the caller to see (method name: [METHOD]) as opposed to just ([METHOD])?

**Pull request merged**